src/solver/mod.rs

```diff
@@ -145,48 +155,39 @@ impl<VS: VersionSet, N: PackageName + Display, D: DependencyProvider<VS, N>> Sol
         Ok(steps)
     }
 
-    /// Adds a clause to the solver and immediately starts watching its literals.
-    fn add_and_watch_clause(&mut self, clause: ClauseState) -> ClauseId {
-        let clause_id = self.clauses.alloc(clause);
-        let clause = &self.clauses[clause_id];
-
-        // Add in requires clause lookup
-        if let &Clause::Requires(solvable_id, version_set_id) = &clause.kind {
-            self.requires_clauses
-                .push((solvable_id, version_set_id, clause_id));
-        }
-
-        // Start watching the literals of the clause
-        let clause = &mut self.clauses[clause_id];
-        if clause.has_watches() {
-            self.watches.start_watching(clause, clause_id);
-        }
-
-        clause_id
-    }
-
     /// Adds clauses for a solvable. These clauses include requirements and constrains on other
     /// solvables.
     ///
     /// Returns the added clauses, and an additional list with conflicting clauses (if any).
     ///
     /// If the provider has requested the solving process to be cancelled, the cancellation value
     /// will be returned as an `Err(...)`.
-    fn add_clauses_for_solvable(
-        &mut self,
+    async fn add_clauses_for_solvable(
+        &self,
         solvable_id: SolvableId,
-    ) -> Result<(Vec<ClauseId>, Vec<ClauseId>), Box<dyn Any>> {
-        if self.clauses_added_for_solvable.contains(&solvable_id) {
-            return Ok((Vec::new(), Vec::new()));
-        }
-
-        let mut new_clauses = Vec::new();
-        let mut conflicting_clauses = Vec::new();
+    ) -> Result<AddClauseOutput, Box<dyn Any>> {
+        let mut output = AddClauseOutput::default();
         let mut queue = vec![solvable_id];
         let mut seen = HashSet::new();
         seen.insert(solvable_id);
 
         while let Some(solvable_id) = queue.pop() {
+            let mutex = {
+                let mut clauses = self.clauses_added_for_solvable.borrow_mut();
+                let mutex = clauses
+                    .entry(solvable_id)
+                    .or_insert_with(|| Rc::new(smol::lock::Mutex::new(false)));
+                mutex.clone()
+            };
+
+            // This prevents concurrent requests to add clauses for a solvable from racing. Only the
+            // first request for that solvable will go through, and others will wait till it
+            // completes.
+            let mut clauses_added = mutex.lock().await;
```

Comment:
Since this await is inside a loop, we could make concurrent metadata fetching even more aggressive by writing this loop in such a way that it processes its items concurrently... I didn't do it, just to keep things simple, but let me know if that's desirable after all!

Reply:
I do think that would be quite nice! But maybe we can do that in another PR?

Reply:
Seems good to me!